relayer/src/ethereum_beacon_client/mod.rs

```diff
@@ -0,0 +1,227 @@
+use std::{
+    cmp::{self, Ordering},
+    error::Error,
+    fmt,
+    time::Duration,
+};
+
+use anyhow::{anyhow, bail, Error as AnyError, Result as AnyResult};
+use checkpoint_light_client_io::{BeaconBlockHeader, Slot};
+use ethereum_common::{
+    beacon::Block as BeaconBlock,
+    utils::{
+        BeaconBlockHeaderResponse, BeaconBlockResponse, FinalityUpdate, FinalityUpdateResponse,
+        UpdateResponse,
+    },
+    MAX_REQUEST_LIGHT_CLIENT_UPDATES,
+};
+use reqwest::{Client, ClientBuilder, RequestBuilder};
+use serde::{de::DeserializeOwned, Deserialize};
+
+pub mod slots_batch;
+pub mod utils;
+
+#[derive(Clone, Debug)]
+pub struct ErrorNotFound;
+
+impl fmt::Display for ErrorNotFound {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt::Display::fmt("Not found (404)", f)
+    }
+}
+
+impl Error for ErrorNotFound {}
+
+#[allow(dead_code)]
+#[derive(Deserialize)]
+struct CodeResponse {
+    code: u64,
+    message: String,
+}
+
+#[derive(Clone)]
+pub struct BeaconClient {
+    client: Client,
+    rpc_url: String,
+}
+
+impl BeaconClient {
+    pub async fn connect(rpc_url: String, timeout: Option<Duration>) -> AnyResult<Self> {
```

Comment:
Sorry but cannot agree with such constructor-method since it doesn't follow the usual Rust-pattern.

Creating methods produce the objects with a guarantee of its existence. Calling any object's method may fail due to the inner method logic but not due to the object existence.

On the other hand we have there a method producing 'a mirror' of the remote 'object'. Constructor may execute successfully then remote endpoint fall down and any subsequent calls to BeaconCLient methods will fail. So the suggestion is to produce a simple `new`:
```rust
    pub fn new(rpc_url: String, timeout: Option<Duration>) -> Self {
        let client = ClientBuilder::new();
        let client = match timeout {
            Some(timeout) => client.timeout(timeout),
            None => client,
        };

        let client = client
            .build()
            .expect("Failed to create reqwest http client");
        Ok(Self { client, rpc_url })
    }
```